### PR TITLE
CSI-1664 Adding resource exhausted status message where needed

### PR DESCRIPTION
**Pull request by zingero**

### CODE REVIEW

File: controller/array_action/array_mediator_ds8k.py [hunk 1]
```diff
@@ -413,6 +414,10 @@ def map_volume(self, volume_id, host_name):
             return lun
         except exceptions.NotFound:
             raise array_errors.HostNotFoundError(host_name)
+        except exceptions.InternalServerError as ex:
+            if ERROR_CODE_MAP_VOLUME_NOT_ENOUGH_EXTENTS in str(ex.message).upper():
+                raise array_errors.NotEnoughSpaceInPool(volume_id)
```
[the remark this answers is not in the document]
Comment (zingero) [reply controller/array_action/array_mediator_ds8k.py]: Done

File: controller/array_action/array_mediator_ds8k.py [hunk 2]
```diff
@@ -413,6 +414,10 @@ def map_volume(self, volume_id, host_name):
             return lun
         except exceptions.NotFound:
             raise array_errors.HostNotFoundError(host_name)
+        except exceptions.InternalServerError as ex:
+            if ERROR_CODE_MAP_VOLUME_NOT_ENOUGH_EXTENTS in str(ex.message).upper():
```
[the remark this answers is not in the document]
Comment (zingero) [reply controller/array_action/array_mediator_ds8k.py:418]: They raise NoAvailableLun/LunAlreadyInUse (CSI-3168)

File: controller/tests/array_action/ds8k/test_array_mediator_ds8k.py [hunk 3]
```diff
@@ -312,6 +312,11 @@ def test_map_volume_volume_not_found(self):
         with self.assertRaises(array_errors.ObjectNotFoundError):
             self.array.map_volume("fake_name", "fake_host")
 
+    def test_map_volume_out_of_space(self):
```
Comment (oriyarde) [inline controller/tests/array_action/ds8k/test_array_mediator_ds8k.py:315]: I don't think this is about space

File: controller/array_action/array_mediator_ds8k.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (oriyarde) [reply controller/array_action/array_mediator_ds8k.py]: I still don't see any change in the mediator interface/other mediators.
from what I see, "not enough extents" is about space, not volume mapping to hosts, so please use a different exception name, e.g. `NoAvailableLunError`

File: controller/tests/array_action/ds8k/test_array_mediator_ds8k.py (hunk 3 above)
Comment (zingero) [reply controller/tests/array_action/ds8k/test_array_mediator_ds8k.py]: Done

File: controller/array_action/array_mediator_ds8k.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (zingero) [reply controller/array_action/array_mediator_ds8k.py]: "Not enough extents" is part of the error message we get from the server. If the server thinks it's related to volume mapping, I think it's OK.

[the remark this answers is not in the document]
Comment (oriyarde) [reply controller/array_action/array_mediator_ds8k.py]: I meant that "extents" doesn't seem to be a cross-storage term, in contrast to "lun"

[the remark this answers is not in the document]
Comment (oriyarde) [reply controller/array_action/array_mediator_ds8k.py]: and that even if we do go with a new exception, it should still be reflected in the mediator interface and the other mediators accordingly

[the remark this answers is not in the document]
Comment (zingero) [reply controller/array_action/array_mediator_ds8k.py]: Done

**Review by ArbelNathan — DISMISSED** [review]
LGTM

File: controller/array_action/array_mediator_xiv.py [hunk 4]
```diff
@@ -341,6 +345,11 @@ def create_snapshot(self, volume_id, snapshot_name, pool=None):
             logger.exception(ex)
             raise array_errors.PermissionDeniedError(
                 "create snapshot {0} from volume {1}".format(snapshot_name, volume_id))
+        except xcli_errors.CommandFailedRuntimeError as ex:
+            logger.exception(ex)
+            if MAXIMUM_SNAPSHOTS_REACHED_ERROR in ex.status:
+                raise array_errors.MaximumSnapshotsReached(pool)
```
Comment (oriyarde) [inline controller/array_action/array_mediator_xiv.py:351]: as I [wrote](https://github.com/IBM/ibm-block-csi-driver/pull/349#discussion_r667476177), new exceptions should be added in the mediator interface and wherever applicable in the mediator implementations

Comment (zingero) [reply controller/array_action/array_mediator_xiv.py:351]: Done

**Review by ArbelNathan — APPROVED** [review]
LGTM

File: controller/array_action/array_mediator_xiv.py [hunk 5]
```diff
@@ -328,6 +328,9 @@ def create_snapshot(self, volume_id, snapshot_name, pool=None):
                                                            vol=source_cli_volume.name).as_single_element
             logger.info("finished creating cli snapshot {0} from volume {1}".format(snapshot_name, volume_id))
             return self._generate_snapshot_response(cli_snapshot)
+        except xcli_errors.MaxVolumesReachedError as ex:
+            logger.exception(ex)
+            raise array_errors.NotEnoughSpaceInPool(volume_id)
```
[the remark this answers is not in the document]
Comment (zingero) [reply controller/array_action/array_mediator_xiv.py]: MaxVolumesReachedError means the pool (maybe) has free space, but the maximum amount of volumes is already reached.

File: controller/array_action/array_mediator_interface.py [hunk 6]
```diff
@@ -245,6 +245,8 @@ def create_snapshot(self, volume_id, snapshot_name, pool=None):
             PermissionDenied
             NotEnoughSpaceInPool
             SnapshotSourcePoolMismatch
+            MaximumSnapshotsReached
+            MaximumVolumesReached
```
Comment (oriyarde) [inline controller/array_action/array_mediator_interface.py:249]: shouldn't `MaximumVolumesReached` also be raised in create volume?

File: controller/array_action/array_mediator_xiv.py (hunk 4 above)
Comment (oriyarde) [reply controller/array_action/array_mediator_xiv.py:351]: what about svc + ds8k?

File: controller/array_action/array_mediator_xiv.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (oriyarde) [reply controller/array_action/array_mediator_xiv.py]: sound strange for create snapshot

File: controller/array_action/array_mediator_xiv.py (hunk 4 above)
Comment (zingero) [reply controller/array_action/array_mediator_xiv.py:351]: Currently, we don't catch similar exceptions in SVC/DS8K. Anyway, it's not related to RE status code. Thus, it's not related to this PR.

File: controller/array_action/array_mediator_interface.py (hunk 6 above)
Comment (zingero) [reply controller/array_action/array_mediator_interface.py:249]: Maybe, but it's not related to RE status code.

File: controller/array_action/array_mediator_xiv.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (zingero) [reply controller/array_action/array_mediator_xiv.py]: I know, but this is a possible error.

File: controller/array_action/array_mediator_xiv.py (hunk 4 above)
Comment (oriyarde) [reply controller/array_action/array_mediator_xiv.py:351]: we shouldn't add new exceptions in the interface without adding them wherever applicable in all the mediators.
so please write the ticket number for the "maximum objects reached" handling and remove it from this PR

Comment (zingero) [reply controller/array_action/array_mediator_xiv.py:351]: "here" == in this conversation or in the code (as a comment)

Comment (oriyarde) [reply controller/array_action/array_mediator_xiv.py:351]: conversation is sufficient (in this case the code doesn't have any smell that requires a ticket comment)

Comment (zingero) [reply controller/array_action/array_mediator_xiv.py:351]: CSI-3183

File: controller/array_action/array_mediator_xiv.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (oriyarde) [reply controller/array_action/array_mediator_xiv.py]: [?](https://github.com/IBM/ibm-block-csi-driver/blob/82b5d058b23206d5f18a5fc57937ede6fa7232b8/controller/array_action/array_mediator_xiv.py#L22)

[the remark this answers is not in the document]
Comment (zingero) [reply controller/array_action/array_mediator_xiv.py]: Done

**Review by oriyarde — APPROVED** [review]
LGTM

**Review by ArbelNathan — APPROVED** [review]
LGTM

**Pull request merged**